confluentinc/cp-demo
# DEVX-2162: add demo artifacts for Self Balancing Clusters (SBC)

### Description 

https://confluentinc.atlassian.net/browse/DEVX-2162

_What behavior does this PR change, and why?_

- Adds configuration and scripting for a third broker, `kafka3` which can join the cluster but is not started by-default.
- The partition count for topics is increased from 2 to 6 to better demonstrate cluster-rebalancing.  
- Self Balancing Cluster configuration for self-healing on broker failure is reduced (for demo) from 1 hour to 30 seconds.
- Scripts are added to start broker `kafka3` and monitor the rebalancing logs, and to kill `kafka3` and monitor the logs while cluster healing is underway.

### Author Validation

_Describe the validation already done, or needs to be done, by the PR submitter._


- [X] Run cp-demo
- [ ] jmx-monitoring-stacks

### How to test

- Start `cp-demo` as normal.
- Once startup has completed, you can either proceed with the SBC add-broker demo immediately, or wait around 15 minutes.  If you start immediately, an add-broker task will register immediately and be visible, but there will be a 15 minute or so delay while SBC collects statistics for the rebalance plan.
- Run `scripts/sbc/add-broker.sh` which will start broker `kafka3` then wait-with-timeout for SBC to log a message that it has created a rebalance task.
- Visit the Control Center UI and review the Self Balancing Cluster summary panel and see an Add Broker task appear.
- Run `docker-compose exec kafka1 kafka-topics   --describe   --topic pageviews --bootstrap-server kafka1:9091   --command-config /etc/kafka/secrets/client_sasl_plain.config` - use `watch` if desired ... look for eventual partitions, leaders and replicas being moved onto broker 3.
- Note Add Broker task completes - check with scripts `scripts/validate/validate_sbc_add_broker_reassignment.sh` and `scripts/validate/validate_sbc_add_broker_completed.sh`.

Test broker removal - option 1, scheduled removal, scale-down

- Visit broker 3 in the Control Center UI and use the Remove Broker option.
- Tail logs: `docker-compose logs -f --tail=1000 | grep -E "(databalancer|cruisecontrol)" | grep -v "confluent.balancer.class"`
- Run `docker-compose exec kafka1 kafka-topics   --describe   --topic pageviews --bootstrap-server kafka1:9091   --command-config /etc/kafka/secrets/client_sasl_plain.config` and note under-replicated partitions.  See these also in the Control Center UI.
- SBC will eventually complete and rebalance the partitions that were on broker 3 across brokers 1 and 2.
- Run `kafka-topics` again and check no URPs remain.

Test broker removal - option 2, unscheduled

- Run `scripts/sbc/kill-broker.sh`, which will stop broker `kafka3` and then watch for SBC to log a message that self-healing has commenced.
- SBC will take 30 seconds to act on lost broker.
- Perform similar monitoring to scheduled removal - URPs should be 0 when rebalancing has completed.
- Validate completion some time later with `scripts/validate/validate_sbc_kill_broker_completed.sh`.

### Reviewer Tasks

_Describe the tasks/validation that the PR submitter is requesting to be done by the reviewer._
